resample/bootstrap.py

```diff
@@ -0,0 +1,285 @@
+
+
+def resample(
+    sample: Sequence,
+    size: int = 1000,
+    method: str = "balanced",
+    strata: Optional[Sequence] = None,
+    random_state: Optional[Union[np.random.Generator, int]] = None,
+) -> np.ndarray:
+    """
+    Generator of bootstrap samples.
+
+    Parameters
+    ----------
+    sample : array-like
+        Original sample.
+    size : int, optional
+        Number of bootstrap samples to generate. Default is 1000.
+    method : str or None, optional
+        How to generate bootstrap samples. Supported are 'ordinary', 'balanced', or
+        a distribution name for a parametric bootstrap. Default is 'balanced'.
+        Supported distribution names: 'normal' (also: 'gaussian', 'norm'),
+        'student' (also: 't'), 'laplace', 'logistic', 'F' (also: 'f'),
+        'beta', 'gamma', 'log-normal' (also: 'lognorm', 'log-gaussian'),
+        'inverse-gaussian' (also: 'invgauss'), 'pareto', 'poisson'.
+    strata : array-like, optional
+        Stratification labels. Must have the same shape as `sample`. Default is None.
+    random_state : numpy.random.Generator or int, optional
+        Random number generator instance. If an integer is passed, seed the numpy
+        default generator with it. Default is to use `numpy.random.default_rng()`.
+
+    Notes
+    -----
+    Stratification:
+
+    If data is not iid, but consists of several distinct classes, stratification
+    ensures that the relative proportions of each class are maintained in each
+    replicated sample. This is a stricter constraint than that offered by the
+    balanced bootstrap, which only guarantees that classes have the original
+    proportions over all replicas.
+
+    Yields
+    ------
+    ndarray
+        Bootstrap sample.
+    """
+    if random_state is None:
+        rng = np.random.default_rng()
+    elif isinstance(random_state, int):
+        rng = np.random.default_rng(random_state)
+    else:
+        rng = random_state
+
+    sample = np.atleast_1d(sample)
+
+    if strata is not None:
+        strata = np.atleast_1d(strata)
+        if strata.shape != sample.shape:  # type: ignore
+            raise ValueError("a and strata must have the same shape")
+        return _resample_stratified(sample, size, method, strata, rng)
+
+    if method == "balanced":
+        return _resample_balanced(sample, size, rng)
+    if method == "ordinary":
+        return _resample_ordinary(sample, size, rng)
+
+    dist = {
+        # put aliases here
+        "gaussian": stats.norm,
+        "normal": stats.norm,
+        "log-normal": stats.lognorm,
+        "log-gaussian": stats.lognorm,
+        "inverse-gaussian": stats.invgauss,
+        "student": stats.t,
+    }.get(method, None)
+
+    if dist is None:
+        # use scipy.stats name
+        try:
+            dist = getattr(stats, method.lower())
+        except AttributeError:
+            raise ValueError("Invalid family: '{}'".format(method))
```

Comment:
```suggestion
            raise ValueError(f"Invalid family: '{method}'")
```

Reply:
I love f-strings, but I thought I cannot use them, because we want to support older Python versions, but it turns out they were already added in 3.6.